src/frontend/src/optimizer/mod.rs

```diff
@@ -429,56 +434,166 @@ impl PlanRoot {
         append_only: bool,
         watermark_descs: Vec<WatermarkDesc>,
         version: Option<TableVersion>,
+        with_external_source: bool,
     ) -> Result<StreamMaterialize> {
-        let mut stream_plan = self.gen_optimized_stream_plan(false)?;
+        let stream_plan = self.gen_optimized_stream_plan(false)?;
 
-        // Add DML node.
-        stream_plan = StreamDml::new(
-            stream_plan,
-            append_only,
-            columns
+        assert!(!pk_column_ids.is_empty() || row_id_index.is_some());
+
+        let pk_column_indices = {
+            let mut id_to_idx = HashMap::new();
+
+            columns.iter().enumerate().for_each(|(idx, c)| {
+                id_to_idx.insert(c.column_id(), idx);
+            });
+            pk_column_ids
                 .iter()
-                .filter(|&c| (!c.is_generated()))
-                .map(|c| c.column_desc.clone())
-                .collect(),
-        )
-        .into();
+                .map(|c| id_to_idx.get(c).copied().unwrap()) // pk column id must exist in table columns.
+                .collect_vec()
+        };
+
+        fn inject_project_if_needed(columns: &[ColumnCatalog], node: PlanRef) -> Result<PlanRef> {
+            let exprs = LogicalSource::derive_output_exprs_from_generated_columns(columns)?;
+            if let Some(exprs) = exprs {
+                let logical_project = generic::Project::new(exprs, node);
+                return Ok(StreamProject::new(logical_project).into());
+            }
+            Ok(node)
+        }
+
+        fn inject_dml_node(
+            columns: &[ColumnCatalog],
+            append_only: bool,
+            stream_plan: PlanRef,
+            pk_column_indices: &[usize],
+            kind: PrimaryKeyKind,
+            column_descs: Vec<ColumnDesc>,
+        ) -> Result<PlanRef> {
+            let mut dml_node = StreamDml::new(stream_plan, append_only, column_descs).into();
+
+            // Add generated columns.
+            dml_node = inject_project_if_needed(columns, dml_node)?;
+
+            dml_node = match kind {
+                PrimaryKeyKind::UserDefinedPrimaryKey | PrimaryKeyKind::RowIdAsPrimaryKey => {
+                    RequiredDist::hash_shard(pk_column_indices)
+                        .enforce_if_not_satisfies(dml_node, &Order::any())?
+                }
+                PrimaryKeyKind::AppendOnly => StreamExchange::new_no_shuffle(dml_node).into(),
+            };
```

Comment:
The dist keys changes is because now we generate exchange here but previouesly we add it in materialize node. with https://github.com/risingwavelabs/risingwave/blob/aa5e7988fe85acb5b4bf06c0340ec30e24627cee/src/frontend/src/optimizer/mod.rs#L482